vscode/src/diagnostics.ts

```diff
@@ -130,3 +140,151 @@ export function updateQSharpJsonDiagnostics(
+// They pertain to the last-run command. They don't get cleared or refreshed when the
+// user edits the document. They will be cleared when a new command is run.
+// They can also be dismissed by the user with a code action.
+//
+
+let commandDiagnostics: vscode.DiagnosticCollection | undefined;
+
+function startCommandDiagnostics(): vscode.Disposable[] {
+  commandDiagnostics =
+    vscode.languages.createDiagnosticCollection(qsharpLanguageId);
+
+  const dismissCommand = vscode.commands.registerCommand(
+    "qsharp-vscode.dismissCommandDiagnostics",
+    clearCommandDiagnostics,
+  );
+
+  const dismissActionProvider = vscode.languages.registerCodeActionsProvider(
+    qsharpLanguageId,
+    {
+      provideCodeActions(doc, range, context): vscode.CodeAction[] | undefined {
+        const commandErrors = context.diagnostics.filter((d) =>
+          d.message.startsWith("Q# command error: "),
+        );
+
+        if (commandErrors.length > 0) {
+          const action = new vscode.CodeAction(
+            "Dismiss errors for the last run Q# command",
+          );
+          action.diagnostics = commandErrors;
+          action.command = {
+            command: "qsharp-vscode.dismissCommandDiagnostics",
+            title: "Dismiss errors for the last run Q# command",
+          };
+          action.isPreferred = true;
+          return [action];
+        }
+      },
+    },
+  );
+
+  return [commandDiagnostics, dismissCommand, dismissActionProvider];
+}
+
+export function clearCommandDiagnostics() {
+  if (!commandDiagnostics) {
+    log.warn(`no diagnostic collection for commands, not clearing`);
+  }
+
+  commandDiagnostics?.clear();
+}
+
+/**
+ * Wrapper that invokes a function. This function would be something that calls into
+ * the Q# compiler. If the function throws an exception that contains Q# diagnostics,
+ * reports the diagnostics to VS Code before rethrowing the original exception.
+ */
+export async function invokeAndReportCommandDiagnostics<T>(
+  fn: () => T | Promise<T>,
+): Promise<T> {
+  try {
+    // Clear the diagnostics from the last command run.
+    clearCommandDiagnostics();
+    return await fn();
+  } catch (e: unknown) {
+    reportIfQSharpErrors(e);
+    throw e;
+  }
+}
+
+/**
+ * Given an exception, checks if it's a JSON string representation of IQSharpError[],
+ * and if so, reports the diagnostics to VS Code.
+ *
+ * @param e an exception originating from the qsharp-lang package
+ */
+function reportIfQSharpErrors(e: unknown) {
+  if (!commandDiagnostics) {
+    log.warn(`diagnostic collection for commands was not initialized`);
+    return;
+  }
+
+  let qsharpErrors: IQSharpError[] | undefined;
+  if (typeof e === "string") {
+    try {
+      const errors = JSON.parse(e);
+      // Check for the shape of IQSharpError[]
+      if (Array.isArray(errors) && errors.length > 0 && errors[0].document) {
+        qsharpErrors = errors;
+      }
+    } catch (_) {
```

Comment:
Should there maybe be a log.warn here or something? Is this expected to happen ever?

Reply:
Right now, no. But there are some functions in the WASM layer that just throw message `string`s as errors. It's conceivable that those may come through this code path in the future. I guess warning in that case would be a good thing... throwing raw strings is not a great idea in general.